lib/execution_engine2/utils/catalog_cache.py

```diff
@@ -0,0 +1,89 @@
+from collections import defaultdict
+from typing import Dict
+
+from lib.installed_clients.CatalogClient import Catalog
+
+
+class CatalogCache:
+    """
+    Per call catalog cache used to speed up catalog lookups
```

Comment:
Also note that the cache is not thread safe

Reply:
What does it mean the cache is not thread safe? multiple requests to the cache have the chance to both update the cache ?

Reply:
It's not designed to work with threads reliably. Imagine this sequence of events:

- thread 1 requests client group info (CGI) for an app from the cache
- thread 2 requests CGI for the same app from the cache
- the catalog returns CGI to thread 1
- thread 1 updates the cache
- the catalog's CGI is updated for the app 
- the catalog returns the new CGI to thread 2
- thread 2 updates the cache

The expected behavior for a cache like this would be that both thread 1 and thread 2 get the same CGI, but in this situation they don't.

Which is fine - thread safety isn't needed for the use case. But we should document that so that someone later doesn't try and use it for a use case it's not designed for.

Reply:
:+1: